lib/protobuf/dsl.ex

```diff
@@ -404,4 +431,43 @@ defmodule Protobuf.DSL do
   defp type_numeric?(:float), do: true
   defp type_numeric?(:double), do: true
   defp type_numeric?(_), do: false
+
+  # We do this because the :| operator is left-associative, so if we just map and build "acc |
+  # spec" then we end up with "((foo | bar) | baz) | bong". By building it from right to left, it
+  # works just fine.
+  defp union_specs(specs) do
+    Enum.reduce(Enum.reverse(specs), fn spec, acc ->
+      quote do: unquote(spec) | unquote(acc)
+    end)
+  end
+
+  # Used by Protobuf.Decoder
+  @doc false
+  def field_default(syntax, field_props)
+
+  def field_default(_syntax, %FieldProps{default: default}) when not is_nil(default), do: default
+  def field_default(_syntax, %FieldProps{repeated?: true}), do: []
+  def field_default(_syntax, %FieldProps{map?: true}), do: %{}
+  def field_default(:proto3, props), do: type_default(props.type)
+  def field_default(_syntax, _props), do: nil
+
+  defp type_default(:int32), do: 0
+  defp type_default(:int64), do: 0
+  defp type_default(:uint32), do: 0
+  defp type_default(:uint64), do: 0
+  defp type_default(:sint32), do: 0
+  defp type_default(:sint64), do: 0
+  defp type_default(:bool), do: false
+  defp type_default({:enum, mod}), do: Code.ensure_loaded(mod) && mod.key(0)
```

Comment:
In the previous implementation this only applied to proto3. Is it safe to generalize? I'm thinking about proto2 enums with non-zero defaults.

Reply:
@britto We still only do this for proto3. `type_default/1` is only called if `syntax == :proto3` up a few lines!